### PR TITLE
Authorization Plugin for Keycloak Identity Server

### LINKED ISSUE #1679: feature : OAuth Authorization plugin for Keycloak
### Issue description
[Keycloak](https://www.keycloak.org/) is an Identity Server from RedHad based on OAuth/OIDC. Keycloak provides a very powerful [policy engine](https://www.keycloak.org/docs/latest/authorization_services/index.html#_policy_overview) for Authorization. It can support Roles based, Attribute-based, and Dynamic authorization strategies. 

The following image shows how APISIX can be integrated with Keycloak Server.

image

The following is a [proposed schema](https://www.keycloak.org/docs/latest/authorization_services/index.html#_service_overview) for the authz-keycloak plugin:
```bash
curl http://127.0.0.1:9080/apisix/admin/routes/5  -X PUT -d '
{
    "uri": "/get",
    "host": "httpbin.org",
    "plugins": {
        "authz-keycloak": {
          "token_endpoint": "token_endpoint_url",
          "grant_type": "urn:ietf:params:oauth:grant-type:uma-ticket",
          "enforcement_policy": "Permissive OR Enforcing",
          "permissions": [{"resourse_name#scope_1_name"}, {{"resourse_name#scope_2_name"}}]
        }
    },
    "upstream_id": 50
}'
```
`Policy enforcing mode` will reject calls that do not have associated permissions and `Permissive enforcement policy` will allow the API call if no permission is tied with the resource/path.

It's not mandatory to provide the permissions but if not provided APISIX should call Keycloak twice to first identify the protected resource path and in the second call to evaluate the permissions.

I would like to contribute to this plugin and would like to hear the thoughts of the community.

### PR DESCRIPTION
**Pull request by sshniro**
Resolves #1679 

This plugin provides granular [access controls (RBAC/ ABAC / JS Based)](https://www.keycloak.org/docs/latest/authorization_services/#_policy_overview) for the routes via the Keycloak Identity Server.

### CODE REVIEW

File: apisix/plugins/authz-keycloak.lua [hunk 1]
```diff
@@ -0,0 +1,128 @@
...
+            ["Content-Type"] = "application/x-www-form-urlencoded",
+            ["Authorization"] = token
+        },
+        keepalive_timeout = 60000,
+        keepalive_pool = 5
+    })
+
+    if not httpc_res then
+        core.response.exit(500, httpc_err)
+        core.log.error("error while sending authz request to [" .. host .. "] port["
```
Comment (membphis) [inline apisix/plugins/authz-keycloak.lua:90]: String concatenation should be avoided here.

that is the good style: `core.log.error("xxx", "bb", "c")`

File: apisix/plugins/authz-keycloak.lua [hunk 2]
```diff
@@ -0,0 +1,128 @@
...
+local sub_str  = string.sub
+local plugin_name = "authz-keycloak"
+local url = require "net.url"
+local tostring = tostring
+
+
+local schema = {
+    type = "object",
+    properties = {
+        token_endpoint = {type = "string"},
```
Comment (membphis) [inline apisix/plugins/authz-keycloak.lua:28]: I think we need to limit his maximum and minimum length.

File: apisix/plugins/authz-keycloak.lua [hunk 3]
```diff
@@ -0,0 +1,128 @@
...
+        permissions = {type = "string"},
```
Comment (membphis) [inline apisix/plugins/authz-keycloak.lua:29]: ditoo ?

File: apisix/plugins/authz-keycloak.lua [hunk 4]
```diff
@@ -0,0 +1,128 @@
...
+        grant_type = {
+            type = "string",
+            default="urn:ietf:params:oauth:grant-type:uma-ticket"
+        },
+        audience = {type = "string"},
+        timeout = {type = "integer", minimum = 1, default = 3},
```
Comment (membphis) [inline apisix/plugins/authz-keycloak.lua:35]: It is recommended to use milliseconds here. So the default value should be  3000 .

File: apisix/plugins/authz-keycloak.lua [hunk 5]
```diff
@@ -0,0 +1,128 @@
...
+    if ((not port) and url_decoded.scheme == "https") then
+        port = 443
+    elseif not port then
+        port = 80
+    end
+
+    local httpc = http.new()
+    local httpc_res, httpc_err = httpc:request_uri(conf.token_endpoint, {
+        method = "POST",
+        body = "grant_type=" .. conf.grant_type .. "&audience="
```
[the remark this answers is not in the document]
Comment (sshniro) [reply apisix/plugins/authz-keycloak.lua]: Resolved

File: apisix/plugins/authz-keycloak.lua [hunk 6]
```diff
@@ -0,0 +1,167 @@
...
+            type = "array",
+            items = {
+                type = "string",
+                minLength = 1, maxLength = 100
+            },
+            uniqueItems = true
+        },
+        grant_type = {
+            type = "string",
+            default="urn:ietf:params:oauth:grant-type:uma-ticket"
```
Comment (membphis) [inline apisix/plugins/authz-keycloak.lua:40]: do we need to limit it with `max_len` and `min_len`?

File: apisix/plugins/authz-keycloak.lua [hunk 7]
```diff
@@ -0,0 +1,167 @@
...
+        },
+        audience = {type = "string"},
```
Comment (membphis) [inline apisix/plugins/authz-keycloak.lua:42]: ditoo

File: apisix/plugins/authz-keycloak.lua [hunk 8]
```diff
@@ -0,0 +1,167 @@
...
+local _M = {
+    version = 0.1,
+    priority = 2000,
+    type = 'auth',
+    name = plugin_name,
+    schema = schema,
+}
+
+function _M.check_schema(conf)
+    local ok, err = core.schema.check(schema, conf)
```
Comment (membphis) [inline apisix/plugins/authz-keycloak.lua:67]: we can use `return core.schema.check(schema, conf)`, it is simpler

File: apisix/plugins/authz-keycloak.lua [hunk 9]
```diff
@@ -0,0 +1,167 @@
...
+    return true
+end
+
+
+local function evaluate_permissions(conf, token)
+    local url_decoded = url.parse(conf.token_endpoint)
+    local host = url_decoded.host
+    local port = url_decoded.port
+
+    if ((not port) and url_decoded.scheme == "https") then
```
Comment (membphis) [inline apisix/plugins/authz-keycloak.lua:89]: bad style, this is better:

```
if not port then
    if ... then
        ...
    else
        ...
    end
end
```

File: apisix/plugins/authz-keycloak.lua [hunk 10]
```diff
@@ -0,0 +1,167 @@
...
+
+    if not is_path_protected(conf) and conf.policy_enforcement_mode == "ENFORCING" then
+        core.response.exit(403)
+        return
+    end
+
+    local httpc = http.new()
+    httpc:set_timeout(conf.timeout)
+
+    local permissions_encoded =  ngx.encode_args({permission = conf.permissions})
```
Comment (membphis) [inline apisix/plugins/authz-keycloak.lua:103]: image

File: apisix/plugins/authz-keycloak.lua [hunk 11]
```diff
@@ -0,0 +1,167 @@
...
+
+function _M.rewrite(conf, ctx)
+    local jwt_token, err = fetch_jwt_token(ctx)
+    if not jwt_token then
+        core.log.error("failed to fetch JWT token: ", err)
+        return 401, {message = "Missing JWT token in request"}
+    end
+
+    evaluate_permissions(conf, jwt_token)
+    core.log.debug("hit keycloak-auth rewrite")
```
Comment (membphis) [inline apisix/plugins/authz-keycloak.lua:163]: we should move this line to No.156

image

File: doc/plugins/authz-keycloak.md [hunk 12]
```diff
@@ -0,0 +1,135 @@
...
+OAuth/OIDC and UMA compliant identity providers as well.
+
+For more information on JWT, refer to [Keycloak Authorization Docs](https://www.keycloak.org/docs/latest/authorization_services) for more information.
+
+## Attributes
+
+|Name           |Requirement    |Description|
+|---------      |--------       |-----------|
+| token_endpoint|required       |A OAuth2-compliant Token Endpoint that supports the urn:ietf:params:oauth:grant-type:uma-ticket grant type.|
+| grant_type    |required       |This parameter is required. Must be urn:ietf:params:oauth:grant-type:uma-ticket.|
```
Comment (membphis) [inline doc/plugins/authz-keycloak.md:44]: is this an optional field?

https://github.com/apache/incubator-apisix/pull/1701/files#diff-3f122a686f16b26c4f875474dd1859bdR54

**Comment by membphis** [comment]
great PR

File: doc/plugins/authz-keycloak.md (hunk 12 above)
Comment (sshniro) [reply doc/plugins/authz-keycloak.md:44]: Yes will update accordingly.

File: apisix/plugins/authz-keycloak.lua (hunk 10 above)
Comment (sshniro) [reply apisix/plugins/authz-keycloak.lua]: Aw, yes my bad. Fixed.

**Comment by sshniro** [comment]
@membphis Thank you for the review, the issues are been fixed now.

File: apisix/plugins/authz-keycloak.lua [hunk 13]
```diff
@@ -0,0 +1,164 @@
...
+            type = "array",
+            items = {
+                type = "string",
+                minLength = 1, maxLength = 100
+            },
+            uniqueItems = true
+        },
+        grant_type = {
+            type = "string",
+            default="urn:ietf:params:oauth:grant-type:uma-ticket",
```
Comment (membphis) [inline apisix/plugins/authz-keycloak.lua:40]: if the `grant_type` is a `required` field, the default value is useless.

Comment (chnliyong) [reply apisix/plugins/authz-keycloak.lua:40]: @sshniro The `grant_type` currently seems can only be `urn:ietf:params:oauth:grant-type:uma-ticket`. Is it better to define **enum** type?

Comment (sshniro) [reply apisix/plugins/authz-keycloak.lua:40]: Yes you are correct, I made it optional, and moved to enums.

**Comment by membphis** [comment]
@sshniro many thx, merged

**Pull request merged**